TOML & Basics

TYWong99 opened the pull request:
This updates the TOML with the directory on my computer, plus the testing file template that I personally use.

TYWong99 committed:
TOML & Basics

This updates the TOML with the directory on my computer, plus the testing file template that I personally use.

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+'''
+Testing file (TY)
+ - I likely will be writing my own test files
+    since my TextEdit crashes sometimes when loading your code... orz
```
That's serious business! Let's take this to trello, tell me what errors you get. All tests should run on both our machines without issue. 

In the meantime, write a more official comment on what this test does.

TYWong99 replied:
I wrote more about the crash on Trello

qtran122 replied:
Ok, I saw your post on trello. We can track that bug there. It doesn't seem to be tied to the actual content of this file though, so let's delete these comments.

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+'''
```
Instead of ''' can we use """ for multi-line comments? That's my preference. Does pylint complain about """?

TYWong99 replied:
I actually prefer ''' since it helps me distinguish the syntax from C#/Java in my head in a way - When I use the " char, my head is automatically tuned into thinking "Oh that's gonna be a string variable". This doesn't happen to me as much with the ' char because it's less often used. ... It's hard to explain, but it's ultimately just my personally preference. If you strongly prefer using """ though, I'm fine with that too.

And well, there's also the lazy reason of not having to hold down the [shift] key when typing the ' char.

qtran122 replied:
Okay, sure. We can keep using '''

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+'''Import'''
+
+import toml
+from logic.common import file_utils as fu
```
Just "from logic.common import file_utils" is fine. No need to rename it to "fu".

Just "fu" by itself is confusing.

Also "fu" is shorthand for an expletive, "eff you"

TYWong99 replied:
Alright I'll keep it long then.

Still, I prefer referring the modules in different names since it means I won't have to scan through the whole file to change the reference if we end up renaming the local files. You _can_ techncially use the "Search & Replace" functionality of most test-editing apps, but that might be error-prone when we have functions named similarly.

Anyway, what's your preferenece on this?

qtran122 replied:
The situation you're referring to is not a problem. For example, say we start with:

`from logic.common import file_utils`

if in the future, we changed file_utils.py to be utils_file, we'd have to edit the top import line to be one of the following:

1. `from logic.common import utils_file`
2. `from logic.common import utils_file as file_utils`

If we go with 2, then we don't have to change the usages of file_utils through the entirety of the file. But the point remains, that we'd have to edit the same line anyway. Therefore, we can make the determination to go with 1 or 2 when the time actually comes.

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+from logic.common import file_utils as fu
+#	from logic.common import level_reader as lr
+#	from logic.common import level_writer as lw
+#	from logic.common import image_exporter as ie
```
Let's remove these comments. These other modules should have their own test suites. And if that test_suite will be here, we'll re-add them when the time comes.

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+#==================================================#
+
+# I put main() inside a function
+#  in order to comment out single lines without affecting indentation
```
??? not sure I get what you're referring to here. Can you elaborate?

TYWong99 replied:
I prefer

```
def main():
#	function1()
	function2()
main()
```

over

```
# function1()
function2()
```

because they still "look" aligned between normal and disabled lines.

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+#  in order to comment out single lines without affecting indentation
+def main():
+	'''The main procedure for each CLI'''
+	fu.start_procedures()
```
single-line comment on what this does. Maybe like so:

`fu.start_procedures() # start the time tracking for program run`

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+	fu.start_procedures()
+
+	# TOML DEMO
+	fu.debug_root("input/root_dir.toml")
```
A test file should do the actual testing. But what's happening here is that file_utils houses the test function, and then the test calls into file_utils to test itself - which is odd. Let's move "debug_root()" into this file.

A unit test for test_file_utils.py would typically look like this:

```
import unittest

class TestFileUtils(unittest.TestCase):

    def test_valid_root_exists(self):
        print("Test 1 - Testing we have a valid root directory)
        root_directory = file_utils.get_first_valid_root("input/root_dir.toml")
        assert_not_equals(None, root_directory)

    def test_root_dir_length(self):
        print("Test 2 - Testing we have 3 root directory contenders)
        root_directories = toml.load("input/root_dir.toml")
        assert_equals(3, len(root_directories))

def main():
    test_valid_root_exists()
    test_root_dir_length()

if __name__ == '__main__':
    unittest.main()
```

Note, I didn't actually test this. But here's the reference for [unit tests](https://docs.python.org/3/library/unittest.html). Can you read up on that and make this file an official unit test following unit test syntax rules?

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+
+	# File I/O DEMO
+	input_path = fu.get_first_valid_root("input/root_dir.toml")
+	input_path += "d30.xml"
```
If we're keeping this part, it should be part of an official unit test. Also, it could have been a one-liner:
`input_path = fu.get_first_valid_root("input/root_dir.toml") + "d30.xml"`

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+'''
+
+#==================================================#
+'''Import'''
```
No need for this comment. We don't need to separate out have a big divider just to highlight our imports

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+'''
+Testing file (TY)
+ - I likely will be writing my own test files
```
I'm down for more test files. Thinking ahead, we should embrace proper Unit Testing procedure and conventions. We'll need to have a test for each tool or module. So let's create a new folder just for tests. (So there'll be three folders at the top-level : input, logic, and tests).

The name convention for these tests will be "test_<whatever_is_being_tested>.py". So in this case, this file will be "test_file_utils.py"

As for the existing "cli_test1.py" - I'll be getting rid of that soon

Reading material for unit tests: https://docs.python.org/3/library/unittest.html

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+		if os.path.isdir(root):
+			return root + "/"
+#	return "ROOT NOT FOUND"
+	return ""
```
instead of returning empty string, let's return None

qtran122 commented on cli_test2.py:
```diff
@@ -0,0 +1,49 @@
+
+main()
+
+#==================================================#
```
Hmmm... I'm not a fan of these giant dividers. There could be something more elegant...
I'll take a more official stance on these soon

TYWong99 replied:
I vaguely remember you can "collapse" an entire function-block or comment-block in many IDE, but since I prefer the simplicity of my notepad-alike, thus I am not able to do these collapsing on my end. To keep my code clean, I choose to use these dividers to divide things into sections more explicitly, which also helps when I'm scrolling through large chunks of code to know which sections I'm currently in.

Anyway, I'm considering writing a simple utility program on my end that can do these conversion automatically for all files within folders within folders etc., e.g. replacing "TAB" with 4 space characters and replacing these giant dividers into blank lines. In a sense, it would kinda be similar to my ColorSwapper that I use before/after uploading/downloading tilesheet - You and I get to work in the different working environments that we're comfortable in, all without needing to changing habits.

qtran122 replied:
That seems like overkill though.

I'm alright with using TAB instead of 4 spaces and will switch over.

For these dividers, instead of ================== are you alright with using ----------------------------- ?

TYWong99 replied:
The utility program is more of a handy all-purpose tool that I would probably use for my other mini-projects too, so don't worry much. Not planning on writing it anytime soon though...

Dividers changed from using '=' to '-'.

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+
+def start_procedures():
+	'''Template starting function'''
+	global time_beg
```
No globals. Make time a class instead.

```
class Clock:

    def __init__(self):
        self.start_time = time.time()
        print("")
        print("\tCommencing procedures...")
        print_linebreak_thick()

    def end_procedures(self):
        '''Template ending function'''
        end_time = time.time()
        time_passed = round((end_time - self.start_time), 3)
        print_linebreak_thick()
        print(f"\t~End of All Procedures~ ({time_passed}s)")
        print("")
```

Usage would be like so:

```
clock = time_utils.Clock() # sets time_beg
... <the inbetween stuff>
clock.end_procedures() # does the time printout
```

TYWong99 replied:
You mentioned before that we should avoid creating new class if tuple can be used instead. From what I'm seeing, the entire purpose of this class is for storing the starting time, whereas the print-functions can be placed in just the utils_file.py instead. Just asking to double-check if making this clock class is the "correct" way to do it.

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+def start_procedures():
+	'''Template starting function'''
+	global time_beg
+	time_beg = time.time()
```
time_beg makes me think of "beg" the verb. Is this a request for time?

Let's go with "start_time" and "end_time". Or "time_start" and "time_end" if you want "time_" as the common prefix

TYWong99 replied:
I like beg because it has exactly the same amount of characters as "end". Kinda the OCD acting up on my end when I want everything to look "aligned" haha

```
time_beg = x
time_end = y
```

versus

```
time_start = x
time_end = y
```

~~EVERYTHING IS BETTER WHEN ALIGNED!!!~~

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+#	return "ROOT NOT FOUND"
+	return ""
+
+def debug_root(root_dir_toml):
```
As mentioned in my other comment, this function should be moved into the test file and made an official unit test

qtran122 commented on logic/common/file_utils.py:
```diff
@@ -0,0 +1,101 @@
+#==================================================#
+'''Print Functions'''
+
+str_linebreak = "=================================================="
```
constants should be all caps:

LINE_BREAK = "=" * 50

But there's no need to have this be a constant. It should just be housed in the function. print_linebreak and print_linebreak_thick can also be combined into one function:

```
def print_line_break(print_thick = False):
    if (print_thick):
        print("\n{0}\n".format("=" * 50))
    else:
        print("-" * 50)

```

usage:
```
print_line_break(True) # prints thick line breaks
print_line_break() # prints thin line breaks
```

Alternatively, you could make it take a number "thickness" argument

TYWong99 replied:
Another personal preference is to put underscore only between "different types". So "print" is the action while "linebreak" is what's being printed, so I personally prefer keeping "linebreak" a single word even though it probably breaks the grammar. Do you have preference on this?

qtran122 reviewed (changes requested):
Alright, 2nd code review. It's smaller and more reasonably sized. I've looked it over and wrote some edit requests. Address them all and it we're good to go!

TYWong99 committed:
Addressing previous comments

I haven't checked what a proper testing unit is yet, but most other comments are tended to here

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,191 @@
+#==================================================#
+'''Print Functions'''
+
+def print_linebreak(upper_spacing = 0, lower_spacing = 0, linebreak_char = '='):
```
let's have the linebreak_char be '-'

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,191 @@
+'''File I/O'''
+
+def check_file_exist(filepath):
+	'''Return False if file is not found, may remove since it's just one-liner'''
```
Yeah, we don't need this function. The wrap around doesn't add any extra clarity, since
os.path.exists(filepath) is super clear already

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,191 @@
+	'''Alternate print function'''
+	print(f"\t{print_str}")
+
+def _print_list(list_s):
```
Hmmm... I'm debating the usefulness of these print functions. Python has a lot of ways to print lists elegantly.

For example, you could use:
`print(' '.join(list_s))`

Or if you want newlines:
`print('\n'.join(list_s))`

Reading material: https://www.geeksforgeeks.org/print-lists-in-python-4-different-ways/#

This function would crash on numbers. Maybe this function can justify its existence by handling both numbers and strings elegantly?

TYWong99 replied:
Hm... I'm not seeing any crash with either my print_list() or your join() methods, do you mind specifying?

qtran122 replied:
Ah, it seems join handles the conversion of numbers to strings, so it's all good.

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,191 @@
+
+
+# DO NOT READ
+# WILL TIDY, SALVAGE, OR REMOVE WHEN NEEDED
```
Noo! Please don't do this. Bad, bad, bad code etiquette. To give you an analogy, it's like if your mom was going out to get some groceries, and she said, "do you need anything"?

And then you asked her to get you a PS5 O_o

Your mom is going out for groceries - not an electronics store. Also, a PS5 is a big ask.

The whole point of branches is that they are cheap and disposable - and there are a lot of them. We can juggle multiple branches at once too. Each branc/pull request addresses one thing. This particular branch was already a bit bloated - don't suddenly introduce other content in the middle of it.

Please get rid of it

TYWong99 replied:
Not too sure about the PS5 analogy, but I get your point. Removed now.

qtran122 reviewed (changes requested):
Looking better - just a few more hanging tasks...

TYWong99 committed:
Fixed + More utils functions

Added more "shared" functions in "utils_file.py", and another testing file for printing & File I/O functions

qtran122 commented on test_unit/test_print.py:
```diff
@@ -0,0 +1,72 @@
+	print(list_s)
+	print('\n'.join(list_s))
+#	utils_file._print_list(list_s)
+	utils_file.print_linebreak()
```
Hmmm... I appreciate the attempt! Though, this isn't how unit tests actually go. The goal of unit tests is that if something breaks in the future due to a future update, the unit tests would catch that and inform you of the failure. So there needs to be actual "assertion" calls (that's a thing that unit tests provide) to check the results against  controlled values.

Since there are no assertion calls here, say a future update makes it so that print_linebreak() actually does nothing. Would this test catch it and inform us with a clear "FAILURE DETECTED" message? It wouldn't.

And that's the advantage of real Unit tests.

Anyway, don't worry about fixing this though, I'll lead with an official unit test check-in in the future so you have an example to work against.

TYWong99 replied:
I still haven't read the Unit Tests that you linked on Trello... orz

But yeah, I can kinda see how they're meant to be used with the little YouTube skit. Regardless, it would be awesome if I'm given a quick example (as opposed to reading bunch of text)! So yeah, if it's not too much of a hassle, I would greatly appreciate some help here since I have 0 experience in this whole Unit Test.

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,106 @@
+#--------------------------------------------------#
+'''String Operations'''
+
+def deflate(list_s, list_end = -1):
```
wait a minute... this seems new? (Again, avoid scope creep). I'll let this one slide - just give it a proper description

qtran122 commented on logic/common/utils_file.py:
```diff
@@ -0,0 +1,106 @@
+	return list_s
+
+def write_file(file_name, list_s):
+	'''Overwrites to the file dircetory with the provided list of string'''
```
*directory

qtran122 reviewed (changes requested):
almost done

TYWong99 committed:
Typo fixed & Comments added
